core/store/src/flat_state.rs

```diff
@@ -142,44 +156,60 @@ mod imp {
                 };
                 // TODO: initialize flat storage states correctly, the current FlatStorageState function
                 // doesn't take any argument
-                let flat_storage_state = {
-                    let mut flat_storage_states =
-                        self.0.flat_storage_states.lock().expect(POISONED_LOCK_ERR);
-                    flat_storage_states
-                        .entry(shard_id)
-                        .or_insert_with(|| FlatStorageState::new())
-                        .clone()
+                let flat_storage_state = match self.get_flat_storage_state_for_shard(shard_id) {
+                    Some(flat_storage_state) => flat_storage_state,
+                    None => {
+                        return None;
+                    }
                 };
-                Some(FlatState { store: self.0.store.clone(), cache, flat_storage_state })
+                Some(FlatState {
+                    store: self.0.store.clone(),
+                    block_hash,
+                    cache,
+                    flat_storage_state,
+                })
             }
         }
 
         pub fn get_flat_storage_state_for_shard(
             &self,
             shard_id: ShardId,
         ) -> Option<FlatStorageState> {
-            let flat_storage_states = self.0.flat_storage_states.lock().expect(POISONED_LOCK_ERR);
-            flat_storage_states.get(&shard_id).cloned()
+            let mut flat_storage_states =
+                self.0.flat_storage_states.lock().expect(POISONED_LOCK_ERR);
+            Some(
+                flat_storage_states
+                    .entry(shard_id)
+                    .or_insert_with(|| FlatStorageState::new(self.0.store.clone(), shard_id))
+                    .clone(),
+            )
         }
     }
 }
 
 #[cfg(not(feature = "protocol_feature_flat_state"))]
 mod imp {
+    use crate::flat_state::FlatStorageState;
+    use crate::{Store, StoreUpdate};
     use near_primitives::hash::CryptoHash;
     use near_primitives::types::ShardId;
 
-    use crate::flat_state::FlatStorageState;
-    use crate::Store;
-
     /// Since this has no variants it can never be instantiated.
     ///
     /// To use flat state enable `protocol_feature_flat_state` cargo feature.
     #[derive(Clone)]
     pub enum FlatState {}
 
     impl FlatState {
-        pub fn get_ref(&self, _root: &CryptoHash, _key: &[u8]) -> ! {
+        pub fn update_head(
```

Comment:
I think you can remove the `update_head` function here since we are using the `update_head` function in `FlatStateStorage` now.